### DISCUSSION TITLE
Styling datapane page with css

### DISCUSSION THREAD

**Discussion by geahuji** [datapane/datapane#60] [Help]
I ran into an issue while trying datapane today, and unfortunately Google was no help.
I am looking to use datapane as the main tool of exporting html reports for my work.

**But** the report is in Hebrew, and therefore has to be styled from right to left.

I’ve encountered this issue before and I know that I can solve this with css, is there a way to add styling css file to the datapane page? I saw that in discussion #42 You mentioned the HTML workaround, but I tried to put a "style" code snippet at the beginning of the code and it didn't work,

(I created a code for example if needed, although the idea itself can be applied to many problems)

```
import datapane as dpa

df1 = pd.DataFrame(np.random.randn(6, 4), index=pd.date_range("20130101", periods=6), columns=list("ABCD"))
df2 = pd.DataFrame(np.random.randn(6, 4), index=pd.date_range("20140101", periods=6), columns=list("EFGH"))

report = dpa.Report(
    
    dpa.Page(label="דו\"ח מצב",
        blocks=[
            dpa.HTML("<h2>הנה דו\"ח לדוגמא:</h2>"),
            dpa.Select(
                dpa.Group(
                    dpa.Text("טבלה לבוגר"),
                    dpa.Table(df1,caption="טבלה לדוגמא"),columns=1),
                dpa.Group(
                    dpa.Text("טבלה למוסמך"),
                    dpa.Table(df2,caption="טבלה שנייה לדוגמא"),columns=1)
            ,type = dpa.SelectType.TABS)
        ]
    ),
    dpa.Page(label="אוכלוסיות מיוחדות",
        blocks=[
            dpa.HTML("<h2>הנה דו\"ח לדוגמא:</h2>"),
            dpa.Select(
                dpa.Group(
                    dpa.Text("טבלה לבוגר - ערבים"),
                    dpa.Table(df1,caption="טבלה לדוגמא"),columns=1),
                dpa.Group(
                    dpa.Text("טבלה למוסמך - צ'רקסים"),
                    dpa.Table(df2,caption="טבלה שנייה לדוגמא"),columns=1)
            ,type = dpa.SelectType.TABS)
        ]
    )
)


report.save(path = os.getcwd()+"\Output"+"\Report_test.html",open = True)
```

**Comment by lanthias** [comment]
Thanks a lot for the issue! That's a great idea. We actually have a feature coming which will allow you to apply custom stylesheets to your entire report. This would fix this, as it would allow you to specify that the text direction should always be rtl using CSS.

I'll keep this updated with that feature, which should be landing soon!

In the meantime, you can always use our HTML block as an escape-hatch to set custom css that enables rtl text, i.e:

```python
html = """
<html>
<p style="direction: rtl">טבלה למוסמך - צ'רקסים</p>
</html
"""
dp.Report(dp.HTML(html)).save(path='test.html', open=True)
```
![image](https://user-images.githubusercontent.com/3541695/103683253-db169d80-4f81-11eb-9332-59ea06fe458e.png)

Or make a reusable function to create a custom block using our HTML block:

```python
def rtl(text):
    return dp.HTML(f"""<html>
<p style="direction: rtl">{text}</p>
</html>
""")

dp.Report(rtl("טבלה למוסמך - צ'רקסים")).save(path='test.html', open=True)
```

**Comment by geahuji** [comment]
@lanthias Thanks Leo!! first of all, that's a very good idea!! but what about the tables? and we are talking about a lot of text and tables.
While this is a workaround, what I would like to do, is add a snippet of css code at the beginning of the page that will look like this:
```
<style>
p {
direction: rtl;
}
<\style>
``` 
Which will be applied over all `p` tags in the page.

**Comment by geahuji** [comment]
Hey @lanthias I have a follow-up to my problem of aligning the HTML output to the right.
Is it possible to get the HTML code of the report as an output? e.g. from this code:
```
df = pd.DataFrame(np.random.randn(6, 4), index=pd.date_range("20130101", periods=6), columns=list("אבגד"))

dpa.Report(dpa.HTML(f"""<body style=direction:rtl>
<h4>שלום<\h4><br>
{dp.Table(df)}<br>
<h2>להתראות</h2>""")).save(path = os.getcwd()+"/Output"+f"/Test.html",open = True)
```

currently it just leaves it as an empty space:
![image](https://user-images.githubusercontent.com/53294013/104836066-b391d000-58b3-11eb-8ebf-13fa46002d41.png)
